### PR TITLE
Fixes various issues with completed pods

### PR DESCRIPTION
**Pull request by trozet**
Changes-Include:
 - During pod sync we allocate IPs of all existing pods, but we should
   ignore completed pods
 - During namespace add we add all of the pod IPs to the ns address set,
   but we should ignore completed pods
 - During processing of delete event for a completed pod, we were trying
   to delete the pod again, which would try to free the IP that was
   previously released and could be in use by another pod. We should
   ignore delete events for completed resources as they would have been
   handled during update.
 - On node add, we add all existing pods on that node back to retry as
   an "add". We should skip completed pods here.

Signed-off-by: Tim Rozet <trozet@redhat.com>

### CODE REVIEW

**Comment by trozet** [comment]
@ricky-rav PTAL

File: go-controller/pkg/ovn/obj_retry.go [hunk 1]
```diff
@@ -1166,6 +1166,12 @@ func (oc *Controller) WatchResource(objectsToRetry *retryObjs) *factory.Handler
 					return
 				}
 				klog.Infof("Delete event received for resource %v %s", objectsToRetry.oType, key)
+				// If object is in terminal state, we would have already deleted it during update.
+				// No reason to attempt to delete it here again.
+				if oc.isObjectInTerminalState(objectsToRetry.oType, obj) {
+					klog.Infof("Ignoring delete event for completed resource %v %s", objectsToRetry.oType, key)
```
Comment (tssurya) [inline go-controller/pkg/ovn/obj_retry.go:1172]: hmm didn't we decide to take into consideration the case where we can see a completed pod being retried for a deletion? As in:
podA is created, podA does its business, it goes to completed state. Update Handler will get the event and try to remove its resources (deleteLogicalPort), now let's say that failed mid-way we are now retrying the deletion of completed pod and then we get a deleted event here: our usual logic is to force a deletion here now immediately, or do we just want to return from here and let the retry do its thing?

File: go-controller/pkg/ovn/namespace.go [hunk 2]
```diff
@@ -596,7 +596,7 @@ func (oc *Controller) createNamespaceAddrSetAllPods(ns string) (addressset.Addre
 	} else {
 		ips = make([]net.IP, 0, len(existingPods))
 		for _, pod := range existingPods {
-			if !pod.Spec.HostNetwork {
+			if !pod.Spec.HostNetwork && !util.PodCompleted(pod) {
```
Comment (tssurya) [inline go-controller/pkg/ovn/namespace.go:599]: please tell me if I am being paranoid here:
same case as below:
podA is created, podA does its business, it goes to completed state. Update Handler will get the event and try to remove its resources (deleteLogicalPort), now let's say that failed mid-way we are now retrying the deletion of completed pod and then we upgrade (oops at a bad timing!), now new masters come up and namespace handlers start before the pod handlers, for a brief time when the pod hasn't actually being cleanup should it be part of a namespace_AS?

Comment (tssurya) [reply go-controller/pkg/ovn/namespace.go:599]: I guess I am being a paranoid now..the whole thing is so rare, just ignore my question.. should be 0 probability of happening.

File: go-controller/pkg/ovn/master.go [hunk 3]
```diff
@@ -1374,6 +1374,9 @@ func (oc *Controller) addUpdateNodeEvent(node *kapi.Node, nSyncs *nodeSyncs) err
 		klog.V(5).Infof("When adding node %s, found %d pods to add to retryPods", node.Name, len(pods.Items))
 		for _, pod := range pods.Items {
 			pod := pod
+			if util.PodCompleted(&pod) {
+				continue
```
Comment (tssurya) [inline go-controller/pkg/ovn/master.go:1378]: unless we were retrying a deletion from previous cluster? or actually I guess we don't have to worry about this since reboots and retries won't go well together? retry handler isn't persistent is it?

File: go-controller/pkg/ovn/pods.go [hunk 4]
```diff
@@ -153,6 +147,16 @@ func (oc *Controller) deleteLogicalPort(pod *kapi.Pod, portInfo *lpInfo) (err er
 		return fmt.Errorf("cannot delete GW Routes for pod %s: %w", podDesc, err)
 	}
 
+	// Releasing IPs needs to happen last so that we can deterministically know that if delete failed that
```
Comment (tssurya) [inline go-controller/pkg/ovn/pods.go:150]: ++

**Review by tssurya — COMMENTED** [review]
overall LGTM, I think you have covered all the buggy portions discovered. I just have some paranoid questions inline, specially around "are retries persistent on reboots?" and "do we care about completed pods that are still stuck on deletion from previous reboot"

thanks for the tests, they were much needed.

File: go-controller/pkg/ovn/pods_test.go [hunk 5]
```diff
@@ -468,6 +468,187 @@ var _ = ginkgo.Describe("OVN Pod Operations", func() {
...
+					"10.128.1.1",
+					"myPod3",
+					"10.128.1.3",
+					"0a:58:0a:80:01:03",
+					namespaceT.Name,
+				)
+				myPod3, err := fakeOvn.fakeClient.KubeClient.CoreV1().Pods(t.namespace).Create(context.TODO(),
+					newPod(t3.namespace, t3.podName, t3.nodeName, t3.podIP), metav1.CreateOptions{})
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
```
Comment (ricky-rav) [inline go-controller/pkg/ovn/pods_test.go:593]: nit: double line! :)

File: go-controller/pkg/ovn/obj_retry.go (hunk 1 above)
Comment (ricky-rav) [reply go-controller/pkg/ovn/obj_retry.go:1172]: Is it important to delete this pod right away or can we wait for `iterateRetryResources` to go over it at the next iteration?

**Comment by tssurya** [comment]
@ricky-rav and I were talking, if we wanted to be more careful/strict to make sure we have fixed allocation of IPs to no more than one pod, we should probably do one of these:

1. (ric's idea): have a map in allocator keeping track of which pod is getting which IP so that when we decide we want to releaseIPs from `deleteLogicalPort` we check this map to see if this IP is indeed tied to this pod.
2. (my idea): let's remove the annotation from the pod completely (or at least update the annotation to make sure the podIPs and macs are nil and/or empty array) when doing deleteLogicalPort and be consistent with addLogicalPort which is adding the annotation. This way if portInfo == nil when we refer to pod.annotations.IPs if its a completed pod it should be a no-op.

Basically find a way to keep release IPs consistent with actual reseved podIPs. - that might be a more proper fix.
NOTE: This PR should fix the bugs we are seeing so these above solutions need not be within the scope of this PR specially if its time sensitive, but just putting this out there (although I think implementing option 2 is not that hard to just update pod annotation if releaseIPs was a success in deleteLogicalPort). wdyt @trozet

File: go-controller/pkg/ovn/namespace.go (hunk 2 above)
Comment (tssurya) [reply go-controller/pkg/ovn/namespace.go:599]: @trozet : can you also exclude notScheduled pods from this list?
I don't know why but it seems like as soon as we start namespace watcher, we get a bunch of pods which are not scheduled and we up with `no pod IPs found` error messages:

```
2022-04-27T22:52:34Z {} W0427 22:50:22.455078       1 namespace.go:645] pod openshift-operator-lifecycle-manager/olm-operator-67b79749bf-pkwmt: no pod IPs found
2022-04-27T22:52:34Z {} I0427 22:50:22.455085       1 namespace.go:222] [openshift-cloud-controller-manager-operator] adding namespace took 13.313983ms
2022-04-27T22:52:34Z {} W0427 22:50:22.455096       1 namespace.go:645] pod openshift-operator-lifecycle-manager/package-server-manager-85dcfddf6b-48mxx: no pod IPs found
2022-04-27T22:52:34Z {} I0427 22:50:22.455098       1 namespace.go:218] [openshift-apiserver-operator] adding namespace
2022-04-27T22:52:34Z {} W0427 22:50:22.455102       1 namespace.go:645] pod openshift-operator-lifecycle-manager/catalog-operator-69b88df49-v7rt7: no pod IPs found
2022-04-27T22:52:34Z {} W0427 22:50:22.455107       1 namespace.go:645] pod openshift-operator-lifecycle-manager/collect-profiles-27518325-hfx99: no pod IPs found
2022-04-27T22:52:34Z {} I0427 22:50:22.455125       1 transact.go:41] Configuring OVN: [{Op:insert Table:Address_Set Row:map[external_ids:{GoMap:map[name:openshift-operator-lifecycle-manager_v4]} name:a13781166516790073860] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:u2596996192}]
2022-04-27T22:52:34Z {} W0427 22:50:22.456874       1 namespace.go:645] pod openshift-marketplace/marketplace-operator-59ffd47c54-grcjm: no pod IPs found
2022-04-27T22:52:34Z {} I0427 22:50:22.456887       1 namespace.go:222] [openshift-operator-lifecycle-manager] adding namespace took 15.129189ms
2022-04-27T22:52:34Z {} I0427 22:50:22.456897       1 namespace.go:218] [kube-public] adding namespace
2022-04-27T22:52:34Z {} I0427 22:50:22.456923       1 transact.go:41] Configuring OVN: [{Op:insert Table:Address_Set Row:map[external_ids:{GoMap:map[name:openshift-marketplace_v4]} name:a10951847637192197900] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:u2596996193}]
2022-04-27T22:52:34Z {} I0427 22:50:22.458737       1 namespace.go:222] [openshift-marketplace] adding namespace took 16.95469ms
2022-04-27T22:52:34Z {} W0427 22:50:22.458741       1 namespace.go:645] pod openshift-kube-scheduler-operator/openshift-kube-scheduler-operator-857f8d5497-dfxlv: no pod IPs found
```

**Review by tssurya — COMMENTED** [review]
Another problem is in add and update handlers, we are always doing something if the pod is a completed pod. So in a span of 3 hrs, each pod gets around 15 update events. Usually for non-completed pods, we don't really call deleteLogicalPort and addLogicaPort for each update event, its controlled. But for completed pod, in both add and update we are blindly calling deleteLogicalPort and this keeps printing "attempting to release IPs" in loop. We need to fix this infinite noise.

File: go-controller/pkg/ovn/pods_test.go (hunk 5 above)
Comment (trozet) [reply go-controller/pkg/ovn/pods_test.go]: fixed!

File: go-controller/pkg/ovn/namespace.go (hunk 2 above)
Comment (trozet) [reply go-controller/pkg/ovn/namespace.go:599]: oh good point... will add that on the next iteration after reviews

File: go-controller/pkg/ovn/obj_retry.go (hunk 1 above)
Comment (trozet) [reply go-controller/pkg/ovn/obj_retry.go:1172]: i think its ok to wait because at this point we have already tried to delete it on a past update event and failed, so now it is following the default backoff retry delete mechanism

**Comment by trozet** [comment]
@ricky-rav @tssurya I tried adding an annotation but then realized that had pitfalls because we may store an "old object", try to delete it and set the annotation, but really we could be setting it on a newer version or different pod altogether with the same name. So the solution I used was to iterate through all non-completed pods on the node and check if the same IP address is being used by any of them. The trick is we could race with another pod handler thread adding a pod and getting the same IP address so I had execute this find with the ip allocator locked. This ensures nothing else is getting an IP while we are searching so we can be sure (or at least as sure as possible) when we examine the current pods from the informer and determine if the IP is in use we are accurate.

Testing downstream here: https://github.com/openshift/ovn-kubernetes/pull/1062

File: go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go [hunk 6]
```diff
@@ -321,6 +321,39 @@ func (manager *LogicalSwitchManager) ReleaseIPs(nodeName string, ipnets []*net.I
...
+	}
+
+	// check if ipam has one of the ip addresses, and then execute the predicate function to determine
+	// if this IP should be released or not
+	for _, ipnet := range ipnets {
+		for _, ipam := range lsi.ipams {
+			cidr := ipam.CIDR()
+			if cidr.Contains(ipnet.IP) {
+				if ipam.Has(ipnet.IP) {
+					return predicate()
```
Comment (ricky-rav) [inline go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:348]: I might be overthinking this... but here we are basically returning the result of `predicate()` as soon as we see that an IP address of this pod is in the node IPAM. `predicate` doesn't take any input arguments and returns false if any of the pod IPs (0 or 1 IPv4 + 0 or 1 IPv6) is found in any other non-completed pods of the same node. 

When it's single stack it all works fine, there's only one IP address to consider, but since the pod might be also dual stack, should we actually pass `ipnet` (that is, the pod IP we're considering) to `predicate`?

For instance, let's say that a our pod A is dual stack, it has IP1 and IP2.
- IP1 (v4): is in node IPAM, is still assigned only to pod A
- IP2 (v6): is in node IPAM, but has been reassigned to another pod.

`ConditionalIPRelease` will process `IP1` first, it will see it is in the node IPAM, it will call `predicate()`; `predicate()` will check *both* IPs of pod A against the IPs of all other pods on the node and will see that IP2 belongs to another pod, so `predicate()` will return false. We were checking if we can release `IP1`, but we got a result for `IP2`. The result is that `IP1` will never be released, unless the other pod that has `IP2` gets deleted at some point.

Is this something to worry about or can we safely ignore it?

Comment (tssurya) [reply go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:348]: this smells like legit problem... we should fix? good catch

File: go-controller/pkg/ovn/pods.go [hunk 7]
```diff
@@ -120,12 +120,53 @@ func (oc *Controller) deleteLogicalPort(pod *kapi.Pod, portInfo *lpInfo) (err er
...
 	}
 
-	var allOps, ops []ovsdb.Operation
-	if ops, err = oc.deletePodFromNamespace(pod.Namespace, podIfAddrs, portUUID); err != nil {
-		return fmt.Errorf("unable to delete pod %s from namespace: %w", podDesc, err)
+	shouldRelease := true
+	// check to make sure no other pods are using this IP before we try to release it if this is a completed pod.
+	if util.PodCompleted(pod) {
+		if shouldRelease, err = oc.lsManager.ConditionalIPRelease(pod.Spec.NodeName, podIfAddrs, func() (bool, error) {
+			pods, err := oc.watchFactory.GetAllPods()
```
Comment (tssurya) [inline go-controller/pkg/ovn/pods.go:127]: is it easy to add `GetAllPodsOnNode(pod.Spec.NodeName)`? Then we don't need to iterate over all pods...although I admit I am not aware of how the PodLister works and if internally effect is the same,,,

File: go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go (hunk 6 above)
Comment (tssurya) [reply go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:348]: oh wait this won't be a problem right? Because Tim's predicate function is actually doing:
```
for _, pAddr := range pAddrs {
	for _, podAddr := range podIfAddrs {
	    if pAddr.Equal(podAddr.IP) {
		klog.Infof("Will not release IP address: %s for pod %s/%s. Detected another pod"+
			" using this IP: %s/%s", pAddr.String(), pod.Namespace, pod.Name, p.Namespace, p.Name)
	        return false, nil
	    }
         }
}
klog.Infof("Releasing IPs for Completed pod: %s/%s, ips: %s", pod.Namespace, pod.Name,
				util.JoinIPNetIPs(podIfAddrs, " "))
return true, nil
```

He sets shouldRelease = true only if all IPs of this pod are not being used by any other pod? Of course the reverse could be a problem where one IPV4 is ready for release but we don't release it because the IPV6 one is held by another pod, but maybe that's ok.

File: go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go [hunk 8]
```diff
@@ -315,16 +313,47 @@ func (manager *LogicalSwitchManager) ReleaseIPs(nodeName string, ipnets []*net.I
...
+				}
+			}
+		}
+	}
+
+	return false, nil
+}
+
```
Comment (tssurya) [inline go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:356]: TODO: add unit tests in logical_switch_manager_test.go for new function (we can do it in FUP since you already have coverage in pods_test.go)

File: go-controller/pkg/ovn/pods.go (hunk 7 above)
Comment (trozet) [reply go-controller/pkg/ovn/pods.go:127]: I tried to do that, but it looked like the API only exposed a way to filter by labels and not the  node.

File: go-controller/pkg/ovn/master.go (hunk 3 above)
Comment (trozet) [reply go-controller/pkg/ovn/master.go:1378]: Doesn't look like we clean up NAT entries, but it shouldn't break anything. Whenever a new pod is added, egress gw/ip is used it should handle the NATs. Something to note for the future though in case we want to clean it up.

File: go-controller/pkg/ovn/namespace.go (hunk 2 above)
Comment (trozet) [reply go-controller/pkg/ovn/namespace.go:599]: done

**Comment by tssurya** [comment]
/lgtm
thanks Tim!

**Comment by tssurya** [comment]
dual stack failures:
```
Apr 29 16:14:07.337: INFO: encountered error during dial (did not find expected responses... 
Tries 39
Command curl -g -q -s 'http://10.244.1.73:9080/dial?request=hostname&protocol=udp&host=fc00:f853:ccd:e793::3&port=30162&tries=1'
retrieved map[]
expected map[netserver-0:{} netserver-1:{} netserver-2:{}])
Apr 29 16:14:07.338: FAIL: failed dialing endpoint, did not find expected responses... 
Tries 39
Command curl -g -q -s 'http://10.244.1.73:9080/dial?request=hostname&protocol=udp&host=fc00:f853:ccd:e793::3&port=30162&tries=1'
retrieved map[]
expected map[netserver-0:{} netserver-1:{} netserver-2:{}]

Full Stack Trace
k8s.io/kubernetes/test/e2e.RunE2ETests(0x23f6d57)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:133 +0x697
k8s.io/kubernetes/test/e2e.TestE2E(0x0)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:136 +0x19
testing.tRunner(0xc000b864e0, 0x7151da0)
	/usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x35a
[AfterEach] [sig-network] [Feature:IPv6DualStack]
  /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:186
Apr 29 16:14:07.338: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "dualstack-9224" for this suite.
```
seem unrelated?

**Comment by trozet** [comment]
yeah I think I've seen this dualstack flake in another PR. Anyway we should look into it.

@andreaskaris maybe you can help me out here since you looked into these before?
https://github.com/ovn-org/ovn-kubernetes/runs/6231186852?check_suite_focus=true

**Comment by trozet** [comment]
Downstream CI looks good, merging
